records/2023-03-28.secret-management.md

```diff
@@ -0,0 +1,189 @@
+# Secret Management Options Analysis
+
+Date: 2023-03-28
+
+## Status
+
+**DRAFT**.
+
+## Context
+
+Currently, secret management is done in several different places depending on the part of the solution. This includes encrypted zip files stored along with the code in GitHub. This is not best practice, and has caused issues with secret rollovers. We are investigating alternative methods to manage secrets that put all secrets in one centrally managed location. 
+
+## Options
+
+### HashiCorp Vault SaaS
+[HashiCorp Vault](https://www.hashicorp.com/products/vault) is an industry standard, cloud agnostic secret management tool. It has integrations with all major cloud providers, platform offerings (Kubernetes/OpenShift), as well as a generic API. There is a SaaS version that has the infrastructure and application patching managed by HashiCorp.
+
+### HashiCorp Vault Self Hosted
+[HashiCorp Vault](https://www.hashicorp.com/products/vault) is an industry standard, cloud agnostic secret management tool. It has integrations with all major cloud providers, platform offerings (Kubernetes/OpenShift), as well as a generic API. There is an open source version that can be self hosted within a Kubernetes Cluster. 
+
+### AWS Secrets Manager
+[AWS Secrets Manager](https://aws.amazon.com/secrets-manager) is a first-party AWS offering that integrates natively with AWS resources, and can be implemented within Kubernetes using a CSI Driver. It is built specifically to manage secrets within AWS.
+
+### AWS Parameter Store
+[AWS Parameter STore](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html) is a first-party AWS offering that integrates natively with AWS resources, and can be implemented within Kubernetes using a CSI Driver. It is built to manage parameters within AWS but is encrypted and could theoretically be used to manage secrets as well.
+
+## Features Comparison
+
+|                     | Fine Grained RBAC | Multiple Secret Stores | Multiple Secret  Backends | Automatic Secret Rotation | Just In Time Authentication | Cloud Agnostic | Managed Service | Private Connection Required | License Cost | People-Time Cost | Infrastructure Cost |
```

Comment:
What would be the advantage for us to use _just in time authentication_? Would that require an API call to the Vault agent or service in places we want this to be populated on a needed basis?

Reply:
In this matrix, could we add some risk estimations/data? Are there options that would guarantee more uptimes? Or would be more secure in general?